File: python/Orderbook.py
from typing import List, Optional
from collections import defaultdict

'''
To-Do:

- implement client class
- hash map to dll and order node secured
- reinstall cancellations and removals
- del hash map values when node is del to prevent memory leaks

'''

'''
class Client:
    def __init__(self, id, exposure, cash) -> None:
        self.clientId: str = id
        self.exposure: int = exposure
        self.cash: float = cash
    
    def trade(self, volume, bidOrAsk, txPrice) -> None:
        self.exposure += (1 if bidOrAsk else -1) * volume
        self.cash += (-1 if bidOrAsk else 1) * txPrice * volume
    
    def getHoldings(self, currPrice) -> float:
        return self.cash + self.currPrice * self.exposure
'''

class Order:
    def __init__(self, orderId, time, bidOrAsk, price, volume, client) -> None:
        self.orderId: int = orderId
        self.datetime: float = time
        self.side: str = bidOrAsk
        self.price: float = price
        self.volume: int = volume
        self.client: str = client

class Node:
    def __init__(self, order) -> None:
        self.item: Order = order
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self, new: Node) -> None:
        self.val = new.item.price
        self.head = new
        self.tail = new

    def exists(self) -> bool:
        return bool(self.head)
  
    def prepend(self, data) -> None:
        new = Node(data)
        new.next = self.head

        if self.head:
            self.head.prev = new
            self.head = new
            new.prev = None
            return
        
        self.head = new
        self.tail = new
        new.prev = None
    
    def append(self, new: Node) -> None:
        # new = Node(data)
        new.prev = self.tail

        if not self.tail:
            self.head = new
            self.tail = new
            new.next = None
            return
        
        self.tail.next = new
        new.next = None
        self.tail = new

    def peekFront(self) -> Optional[Order]:
        if self.head: return self.head.item
        return

    def peekBack(self) -> Optional[Order]:
        if self.tail: return self.tail.item
        return
    
    def popleft(self) -> Optional[Order]:
        if not self.head:
            return
        
        if not self.head.next:
            self.head = None
            self.tail = None
            return
        
        temp = self.head
        temp.next.prev = None
        self.head = temp.next
        temp.next = None
        return temp.item
  
    def popright(self) -> Optional[Order]:
        if not self.tail:
            return
        
        if not self.tail.prev:
            self.head = None
            self.tail = None
            return
        
        temp = self.tail
        temp.prev.next = None
        self.tail = temp.prev
        temp.prev = None
        return temp.item

    def remove(self, target) -> None:
        if not self.head or not target:
            return
        
        if self.head == target:
            self.popleft()
        
        if self.tail == target:
            self.popright()

        if target.next:
            target.next.prev = target.prev
        
        if target.prev:
            target.prev.next = target.next
        
        return

    def printDLL(self) -> None:
        if not self.head:
            return

        path = []

        curr = self.head
        while curr:
            path.append(curr.item.price)
            curr = curr.next
        
        print(path)
        return

class Heap:
    def __init__(self) -> None:
        self.heap: List[List[Order]] = []
    
    def _parent(self, i: int): return (i-1)//2
    def _left(self, i: int): return (2*i + 1)
    def _right(self, i: int): return 2*i + 2

    # Helper function to get 'value' of queue at said index
    def _peek(self, i: int): return self.heap[i].val

    def getMin(self) -> Optional[Order]:
        return self.heap[0].head.item if self.heap and self.heap[0].head else None
    
    def getSize(self) -> int:
        return len(self.heap)
    
    def exists(self) -> bool:
        return True if self.heap and self.heap[0].head else False

    def printHeap(self) -> None:
        out = []
        for h in self.heap:
            lvl = []
            curr = h.head
            while curr:
                lvl.append(curr.item.orderId)
                curr = curr.next
            out.append(lvl)
        print(out)
        return


class MinHeap(Heap):
    def __init__(self) -> None:
        super().__init__()

    def heapify(self, i: int) -> None:
        l: int = self._left(i)
        r: int = self._right(i)
        smallest: int = i

        if l < len(self.heap) and self._peek(l) < self._peek(i):
            smallest = l
        if r < len(self.heap) and self._peek(r) < self._peek(smallest):
            smallest = r
        if smallest != i:
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self.heapify(smallest)
    
    # Insertion does change, takes in a boolean from hashmap to append to queue else, proceeds similarly
    def insert(self, dll: DoublyLinkedList) -> None:
        self.heap.append(dll)
        i: int = len(self.heap) - 1
        while i != 0 and self._peek(self._parent(i)) > self._peek(i):
            self.heap[i], self.heap[self._parent(i)] = self.heap[self._parent(i)], self.heap[i]
            i = self._parent(i)

#     # popOrder rules delete min obsolete
#     def popOrder(self) -> None:
#         if not self.heap:
#             return

#         last = self.heap[0].popleft()

#         if not self.heap[0].head:
#             a, b = self.heap[0], self.heap[-1]
#             self.heap[0], self.heap[-1] = b, a
# #             self.heap[0] = self.heap[-1]

#             self.heap.pop()
#             self.heapify(0)
#         return

    def remove(self, target: DoublyLinkedList) -> None:
        # O (n)
        if not self.heap:
            return
        for i in range(len(self.heap)):
            if self.heap[i] == target:
                break
#         print(self.heap)
        self.heap[i] = self.heap[-1]
#         temp = self.heap[-1]
#         target = copy.deepcopy(self.heap[-1])
#         print(self.heap)
#         target = self.heap[-1]
        self.heap.pop()
        self.heapify(0)

class MaxHeap(Heap):
    def __init__(self) -> None:
        super().__init__()

    def heapify(self, i: int) -> None:
        l: int = self._left(i)
        r: int = self._right(i)
        largest: int = i

        # self.loc[self._peek(i)] = i

        if l < len(self.heap) and self._peek(l) > self._peek(i):
            largest = l
        if r < len(self.heap) and self._peek(r) > self._peek(largest):
            largest = r
        if largest != i:
            self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
            self.heapify(largest)
    
    
    # Insertion does change, takes in a boolean from hashmap to append to queue else, proceeds similarly
    def insert(self, dll: DoublyLinkedList) -> None:
        self.heap.append(dll)
        i: int = len(self.heap) - 1
        while i != 0 and self._peek(self._parent(i)) < self._peek(i):
            self.heap[i], self.heap[self._parent(i)] = self.heap[self._parent(i)], self.heap[i]
            i = self._parent(i)


    # popOrder rules delete min obsolete
#     def popOrder(self) -> None:
#         if not self.heap:
#             return

#         last = self.heap[0].popleft()

#         if not self.heap[0].head:
#             a, b = self.heap[0], self.heap[-1]
#             self.heap[0], self.heap[-1] = b, a
# #             self.heap[0] = self.heap[-1]

#             self.heap.pop()
#             self.heapify(0)
#         return

    def remove(self, target: DoublyLinkedList) -> None:
        # O (n)
        if not self.heap:
            return
        
        for i in range(len(self.heap)):
            if self.heap[i] == target:
                break
#         print(self.heap)
        self.heap[i] = self.heap[-1]
#         temp = self.heap[-1]
#         target = copy.deepcopy(self.heap[-1])
#         print(self.heap)
#         target = self.heap[-1]
        self.heap.pop()
        self.heapify(0)

class OrderBook:
    def __init__(self) -> None:
        self.bestBid: MaxHeap = MaxHeap() # max heap
        self.bestAsk: MinHeap = MinHeap() # min heap
        self.orderMap: dict = {}
        self.volumeMap: dict = defaultdict(dict) # price // side // vol
        self.queueMap: dict = defaultdict(dict) # price // side // queue #DONT UNDERSTAND
        self.txct: int = 0

    
    def _crossedTrade(self, book, order) -> bool:
        if book.exists():
            if order.side == 'BID':
                if order.price >= book._peek(0): 
                    return True
            else:
                if order.price <= book._peek(0): 
                    return True
        return False


    def placeOrder(self, order: Order) -> None:
        oppBook = self.bestAsk if order.side == 'BID' else self.bestBid
        sameBook = self.bestBid if order.side == 'BID' else self.bestAsk

#         self.orderMap[order.orderId] = Node(order)
# 
        while order.volume > 0 and self._crossedTrade(oppBook, order):
#             print(oppBook.printHeap())
            
            matchedOrder = oppBook.getMin()

            self.txct += 1

            txPrice, txVolume = matchedOrder.price, min(order.volume, matchedOrder.volume) # add client side interaction here
            
            matchedOrder.volume -= txVolume
            order.volume -= txVolume

            self.volumeMap[txPrice][matchedOrder.side] -= txVolume

            if matchedOrder.volume == 0: 
                self.cancelOrder(matchedOrder.orderId)
#                 oppBook.popOrder()
#                 del self.orderMap[matchedOrder.orderId]

        if order.volume > 0:
            self._placeResting(order, sameBook)
    
    def _placeResting(self, order, book) -> None: # Figure out the deal with queue map and heap
        
        self.orderMap[order.orderId] = Node(order)

        if order.price not in self.queueMap or order.side not in self.queueMap[order.price]:
            self.queueMap[order.price][order.side] = DoublyLinkedList(self.orderMap[order.orderId]) # [order] # DoublyLinkedList(order)
            book.insert(self.queueMap[order.price][order.side])
            self.volumeMap[order.price][order.side] = order.volume
        else:
            self.queueMap[order.price][order.side].append(self.orderMap[order.orderId]) # DLL
            self.volumeMap[order.price][order.side] += order.volume
    
    def cancelOrder(self, orderId) -> None:
        if orderId in self.orderMap:
            orderNode = self.orderMap[orderId]
            order = orderNode.item
            priceQueue = self.queueMap[order.price][order.side]
            priceQueue.remove(orderNode)
            if not priceQueue.exists():
#                 print('endlvl')
                sameBook = self.bestBid if order.side == 'BID' else self.bestAsk
#                 if order.side == 'BUY':
#                     sameBook.printHeap()
#                     print('new')
                sameBook.remove(priceQueue) # add O(log n) heap removal DONE
#                 if order.side == 'BUY':
#                     sameBook.printHeap()
#                 sameBook.remove(self.queueMap[order.price][order.side]) # add O(log n) heap removal DONE
                del self.queueMap[order.price][order.side]
            self.volumeMap[order.price][order.side] -= order.volume
            del self.orderMap[orderId]
    
    def executeOrder(self, orderId) -> None:
        '''
        To do: add client side handling to update balances
        '''
        self.cancelOrder(orderId)
    
    def partialCancel(self, orderId, volume) -> None:
        if orderId in self.orderMap:
            order = self.orderMap[orderId].item
            order.volume -= volume
            self.volumeMap[order.price][order.side] -= order.volume

    def getVolumeAtPrice(self, price, side):
        return self.volumeMap[price][side]

File: cpp/orderbook/orderbook.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <unordered_map>
#include <algorithm> // for std::min
using namespace std;

class Order {
    public:
        int orderId;
        double datetime;
        string side;
        double price;
        int volume;
        string client;

        Order(int orderId, double datetime, std::string side, double price, int volume, std::string client)
            : orderId(orderId), datetime(datetime), side(side), price(price), volume(volume), client(client) {
        }
};

class Node {
    public:
        Order* item;
        Node* prev;
        Node* next;

        Node(Order* item) : item(item), prev(NULL), next(NULL) {};
};

class DoublyLinkedList {
    public:
        Node* head;
        Node* tail;
        int val;

        DoublyLinkedList(Node* init) : head(init), tail(init), val(init->item->price) {}

        bool exists() {
            bool result = (head) ? true : false;
            return result;
        }

        void prepend(Node* newNode) {
            newNode->next = head;

            if (!head) {
                head = newNode;
                tail = newNode;
                newNode->prev = nullptr;
                return;
            } else {
                head->prev = newNode;
                newNode->prev = nullptr;
                head = newNode;
            }
        }

        void append(Node* newNode) {
            newNode->prev = tail;

            if (!tail) {
                tail = newNode;
                head = newNode;
                newNode->next = nullptr;
                return;
            } else {
                tail->next = newNode;
                newNode->next = nullptr;
                tail = newNode;
            }
        }

        int peekFront() {
            if (head) {
                return head->item->price;
            }
            return -1;
        }

        int peekBack() {
            if (tail) {
                return tail->item->price;
            }
            return -1;
        }

        int popleft() {
            if (!head) {
                return -1;
            }

            Node* temp = head;
            int result = temp->item->price;

            if (!head->next) {
                head = nullptr;
                tail = nullptr;
            } else { 
                temp->next->prev = nullptr;
                head = temp->next;
                temp->next = nullptr;
            }

            delete temp;

            return result;
        }

        // swap to order
        int popright() {
            if (!tail) {
                return -1; // swap for nullptr
            }

            Node* temp = tail;
            int result = temp->item->price;

            if (!tail->prev) {
                head = nullptr;
                tail = nullptr;
            } else {
                tail->prev->next = nullptr;
                tail = temp->prev;
                temp->prev = nullptr;
            }

            delete temp;

            return result;
        }

        void remove(Node* target) {
            if (!head && !tail) {
                return;
            } else if (head->item->orderId == target->item->orderId) {
                popleft();
            } else if (tail->item->orderId == target->item->orderId) {
                popright();
            } else {
                if (target->next) {
                    target->next->prev = target->prev;
                }
                if (target->prev) {
                    target->prev->next = target->next;
                }
            }
        }

        void printDLL() {
            Node* curr = head;
            while (curr != NULL) {
                cout << curr->item->price << " ";
                curr = curr->next;
            }
            cout << endl;
        }
};
class Heap {
    public:
        vector<DoublyLinkedList*> heap;
        int parent(int i) { return (i-1)/2; }
        int left(int i) {return (2*i + 1);}
        int right(int i) {return (2*i + 2);}
        
        int peek(int i) {
            return heap[i]->val;
        }
        bool exists() {return !heap.empty();}
        Order* getMin(){ 
            if(exists() and heap.front()->head) {
                return heap.front()->head->item;
            }
            return nullptr;
        }
        virtual void insert(DoublyLinkedList* k) {return;}
        virtual void popOrder() {return;}
        virtual void remove(DoublyLinkedList* target) {return;}

};

class MinHeap: public Heap {

    void heapify(int i){
        int l = left(i);
        int r = right(i);
        int smallest = i;
        if(l < heap.size() && peek(l) < peek(i))
            smallest = l;
        if(r < heap.size() && peek(r) < peek(smallest))
            smallest = r;
        if(smallest != i){
            swap(heap[i], heap[smallest]);
            heapify(smallest);
        }
    }
public:
    int getSize() {return sizeof(heap);}
    
    void insert(DoublyLinkedList* k) override{
        heap.push_back(k);
        int i = heap.size() - 1;
        while(i != 0 && peek(parent(i)) > peek(i)){
            swap(heap[i], heap[parent(i)]);
            i = parent(i);
        }
    }
    
    void popOrder() override{
        if(heap.size() == 0){
            cout << "Heap is empty" << endl;
            return;
        }
        heap[0] = heap.back();
        // DoublyLinkedList* removedPtr = heap.back();
        heap.pop_back();
        heapify(0);
        // delete removedPtr;
    }
    

    
    void printHeap() {
        for(int i = 0; i < heap.size(); ++i)
            cout << peek(i) << "\t";
        cout << endl;
    }

    void remove(DoublyLinkedList* target) override {
        if (getSize() == 0) {
            return;
        }
        int loc;
        for (int i = 0; i < getSize(); i++) { 
            if (heap[i] == target) {
                loc = i;
                break;
            }
        }
        heap[loc] = heap.back();
        heap.pop_back();
        heapify(loc);
        
        delete target;
    }
    
};

class MaxHeap: public Heap {
    
    void heapify(int i){
        int l = left(i);
        int r = right(i);
        int largest = i;
        if(l < heap.size() && peek(l) > peek(i))
            largest = l;
        if(r < heap.size() && peek(r) > peek(largest))
            largest = r;
        if(largest != i){
            swap(heap[i], heap[largest]);
            heapify(largest);
        }
    }
public:
    int getSize() {return sizeof(heap);}
    
    void insert(DoublyLinkedList* k) override{
        heap.push_back(k);
        int i = heap.size() - 1;
        while(i != 0 && peek(parent(i)) < peek(i)){
            std::swap(heap[i], heap[parent(i)]);
            i = parent(i);
        }
    }
    
    void popOrder() override{
        if(heap.size() == 0){
            cout << "Heap is empty" << endl;
            return;
        }
        heap[0] = heap.back();
        // DoublyLinkedList* removedPtr = heap.back();
        heap.pop_back();
        heapify(0);

        // delete removedPtr;
    }
    
    
    void printHeap() {
        for(int i = 0; i < heap.size(); ++i)
            cout << peek(i) << "\t";
        cout << endl;
    }

    void remove(DoublyLinkedList* target) override {
        if (getSize() == 0) {
            return;
        }
        int loc;
        for (int i = 0; i < getSize(); i++) { 
            if (heap[i] == target) {
                loc = i;
                break;
            }
        }
        heap[loc] = heap.back();
        heap.pop_back();
        heapify(loc);

        delete target;
    }
    
};

class OrderBook {
    private:
        MaxHeap bestBid;
        MinHeap bestAsk;
        unordered_map<int, Node*> orderMap; // orderid // orderref
        unordered_map<int, unordered_map<string, int> > volumeMap; // price // side // volume
        unordered_map<int, unordered_map<string, DoublyLinkedList*> > queueMap; // price // side // dll
        unordered_map<string, unordered_map<string, double> > clientMap; // client // exposure/cash // value
        int txct; // tx count

        bool crossedTrade(Heap* book, Order order) {
            if (book->exists()) {
                cout << book->getMin() << endl;
                if (order.side == "BID") {
                    if (order.price >= book->peek(0)) {
                        return true;
                    }
                } else {
                    if (order.price <= book->peek(0)) {
                        return true;
                    }
                }
            }
            return false;
        }

        void updateHoldings(Order* order, string matchClient, int txVolume, double txPrice) {
            if (order->side == "BID") {
                clientMap[order->client]["Exposure"] += txVolume;
                clientMap[order->client]["Cash"] -= txVolume * txPrice;
                clientMap[matchClient]["Exposure"] -= txVolume;
                clientMap[matchClient]["Cash"] += txVolume * txPrice;
            } else {
                clientMap[order->client]["Exposure"] -= txVolume;
                clientMap[order->client]["Cash"] += txVolume * txPrice;
                clientMap[matchClient]["Exposure"] += txVolume;
                clientMap[matchClient]["Cash"] -= txVolume * txPrice;
            }
        };

        void placeResting(Order order, Heap* book) {
            Node* n1 = new Node(&order);    
            orderMap[order.orderId] = n1;

            if (queueMap.count(order.price) == 0 || queueMap[order.price].count(order.side) == 0) {
                // queueMap[order.price][order.side] = new DoublyLinkedList(orderMap[order.orderId])
                queueMap[order.price][order.side] = new DoublyLinkedList(n1);
                book->insert(queueMap[order.price][order.side]);
                volumeMap[order.price][order.side] = order.volume;
            } else {
                // queueMap[order.price][order.side]->append(orderMap[order.orderId])
                queueMap[order.price][order.side]->append(n1);
                volumeMap[order.price][order.side] += order.volume;
            }

        };

    public:

        void placeOrder(Order order) {
            Heap* oppBook = nullptr;
            Heap* sameBook = nullptr;
            if (order.side == "BID") {
                oppBook = &bestAsk;
                sameBook = &bestBid;
            } else {
                sameBook = &bestAsk;
                oppBook = &bestBid;
            }
            cout << "init books" << endl;
            cout << sameBook << endl;
            cout << oppBook << endl;
            bestAsk.printHeap();
            // cout << oppBook->heap << endl;
            // cout << oppBook->exists() << endl;
            if (clientMap.count(order.client) == 0) {
                clientMap[order.client]["Cash"] = 0;
                clientMap[order.client]["Exposure"] = 0;
            }
            cout << "a" << endl;
            cout << oppBook->exists() << endl;
            cout << order.client << endl;
            cout << "w" << endl;
            // cout << oppBook->peek(0) << endl;
            // cout << crossedTrade(oppBook, order) << endl;
            cout << "b" << endl;
            while (order.volume > 0 && crossedTrade(oppBook, order)) {
                Order* matchedOrder = oppBook->getMin();
                double txPrice = matchedOrder->price;
                int txVolume = min(order.volume, matchedOrder->volume);
                txct += txVolume;
                matchedOrder->volume -= txVolume;
                order.volume -= txVolume;
                updateHoldings(&order, matchedOrder->client, txVolume, txPrice);
                volumeMap[txPrice][matchedOrder->side] -= txVolume;
                if (matchedOrder->volume == 0) {
                    cancelOrder(matchedOrder->orderId);
                    // oppBook->popOrder();
                }
            }
            if (order.volume > 0) {
                placeResting(order, sameBook);
            }
            // new code
            // delete oppBook;
            // delete sameBook;
        };

        void cancelOrder(int orderId) { // prob could use order pointer
            if (orderMap.count(orderId) > 0) {
                Node* orderNode = orderMap[orderId];
                Order* order = orderNode->item;
                DoublyLinkedList* priceQueue = queueMap[order->price][order->side];
                priceQueue->remove(orderNode);
                if (!priceQueue->exists()) {
                    Heap* sameBook = nullptr;
                    if (order->side == "BID") {
                        sameBook = &bestBid;
                    } else {
                        sameBook = &bestAsk;
                    }
                    sameBook->remove(priceQueue);
                    queueMap[order->price].erase(order->side);
                    // new code
                    delete priceQueue;
                }
                volumeMap[order->price][order->side] -= order->volume;
                orderMap.erase(orderId);
                // new code
                delete orderNode->item;
                delete orderNode;
            }
        }; 

        void executeOrder(int orderId, string client) {
            if (orderMap.count(orderId) > 0) {
                Order* order = orderMap[orderId]->item;

                if (clientMap.count(client) == 0) {
                    clientMap[client]["Cash"] = 0;
                    clientMap[client]["Exposure"] = 0;
                }

                updateHoldings(order, client, order->volume, order->price);
                cancelOrder(orderId);
            }
        };

        void partialCancel(int orderId, int vol) {
            if (orderMap.count(orderId) > 0) {
                Order* order = orderMap[orderId]->item;
                order->volume -= vol;
                volumeMap[order->price][order->side] -= vol;
            }
        };
        
        int getVolumeAtPrice(int price, string side) {
            return volumeMap[price][side];
        };

        OrderBook(): txct(0) {};
};



// int main() {
//     // Insert some elements into the heap
//     MaxHeap minHeap;

//     Order* o1 = new Order(1, 12345.6789, "Buy", 1010.0f, 10, "Client A");   
//     Order* o2 = new Order(2, 98765.4321, "Sell", 910.0f, 20, "Client B");
//     Order* o3 = new Order(3, 98765.4321, "Sell", 920.0f, 20, "Client B");
//     Order* o4 = new Order(4, 98765.4321, "Sell", 930.0f, 20, "Client B");

//     Node* n1 = new Node(o1);
//     Node* n2 = new Node(o2);
//     Node* n3 = new Node(o3);
//     Node* n6 = new Node(o4);

//     DoublyLinkedList* node1 = new DoublyLinkedList(n1);
//     DoublyLinkedList* node2 = new DoublyLinkedList(n2);
//     DoublyLinkedList* node3 = new DoublyLinkedList(n3);
//     DoublyLinkedList* node4 = new DoublyLinkedList(n6);

//     minHeap.insert(node1);
//     minHeap.insert(node2);
//     minHeap.insert(node3);
//     minHeap.insert(node4);

//     cout << "Min element: " << minHeap.getMin() << endl;
//     cout << "Heap elements: ";
//     minHeap.printHeap();  // The elements may not be in sorted order after insertion

//     // Remove the minimum element
//     minHeap.popOrder();
//     cout << "After popping the minimum element: ";
//     minHeap.printHeap();  // The heap will be restructured after the pop

//     // Remove a specific element
//     minHeap.remove(node2);
//     cout << "After removing node2: ";
//     minHeap.printHeap();

//     // Clean up memory
//     // delete node1;
//     // delete node2;
//     // delete node3;
//     // delete node4;

//     return 0;
// }
// int main() {
//  // Create a linked list with elements
//     Order* o1 = new Order(1, 12345.6789, "Buy", 1010.0f, 10, "Client A");
//     Order* o2 = new Order(2, 98765.4321, "Sell", 910.0f, 20, "Client B");
//     Order* o3 = new Order(3, 98765.4321, "Sell", 920.0f, 20, "Client B");
//     Order* o4 = new Order(4, 98765.4321, "Sell", 930.0f, 20, "Client B");

//     Node* node1 = new Node(o1);
//     Node* node2 = new Node(o2);
//     Node* node3 = new Node(o3);
//     Node* node6 = new Node(o4);
//     // Node* node7 = new Node(7);

//     DoublyLinkedList list(node1);
//     list.append(node2);
//     list.append(node3);
//     list.append(node6);
//     // list.append(node7);

//     // Display initial list
//     std::cout << "Initial List: ";
//     list.printDLL();

//     // Test Case: Perform popleft
//     int removedValue = list.popleft();
//     std::cout << "Removed Value: " << removedValue << std::endl;

//     // Display updated list
//     std::cout << "Updated List: ";
//     list.printDLL();

//     // Test Case: Perform popright
//     int newval = list.popright();
//     std::cout << "Removed Value: " << newval << std::endl;

//     // Display updated list
//     std::cout << "Updated List: ";
//     list.printDLL();

//     list.remove(node3);

//     // Display updated list
//     std::cout << "Updated List: ";
//     list.printDLL();

//     return 0;
// }
